src/main/kotlin/io/nais/security/oauth2/token/JwkSetFailover.kt

```diff
@@ -0,0 +1,89 @@
+package io.nais.security.oauth2.token
+
+import com.nimbusds.jose.KeySourceException
+import com.nimbusds.jose.jwk.JWK
+import com.nimbusds.jose.jwk.JWKSelector
+import com.nimbusds.jose.jwk.JWKSet
+import com.nimbusds.jose.jwk.source.JWKSource
+import com.nimbusds.jose.proc.SecurityContext
+import com.nimbusds.jose.util.Resource
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import mu.KotlinLogging
+import java.io.IOException
+import java.net.URL
+
+private val log = KotlinLogging.logger {}
+
+class JwkSetFailover(
+    initialKeySource: String,
+    private val jwkSetUri: String,
+    private val options: FailoverOptions
+) : JWKSource<SecurityContext> {
+    private var jwkSet = initialKeySource.toJwkSet()
+    private val mutex = Mutex()
+
+    //  fine-grained Thread safe operation
+    private fun setJWKSet(inputJwks: JWKSet) {
+        runBlocking {
+            mutex.withLock {
+                jwkSet = inputJwks
+            }
+        }
+    }
+
+    fun getJwkSet(): JWKSet? {
+        return this.jwkSet
+    }
+
+    @Throws(KeySourceException::class)
+    override fun get(jwkSelector: JWKSelector, context: SecurityContext?): MutableList<JWK> {
+        try {
+            options.coroutineScope.launch { updateJwkSetResourceAsync() }
+        } catch (t: Throwable) {
+            throw KeySourceException("trying to get current jwks from resource: $jwkSetUri", t)
+        }
+        log.info("failover jwkSet launched")
+        return jwkSelector.select(getJwkSet())
```

Comment:
What happens here if `getJwkSet()` is empty (i.e. the non-blocking coroutine above hasn't set or updated the `jwkSet` variable yet)?